velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -135,79 +125,76 @@ struct RegexpReplaceFunction {
-    size_t utf8Position = 0;
-    size_t numCodePoints = 0;
-    while (numCodePoints < position - 1 && utf8Position <= stringInput.size()) {
-      int charLength =
-          utf8proc_char_length(inputStringPiece.data() + utf8Position);
-      VELOX_USER_CHECK_GT(
-          charLength, 0, "regexp_replace encountered invalid UTF-8 character");
-      ++numCodePoints;
-      utf8Position += charLength;
-    }
     if (utf8Position > stringInput.size() + 1) {
-      result.resize(inputStringPiece.size());
-      std::memcpy(
-          result.data(), inputStringPiece.data(), inputStringPiece.size());
+      result = stringInput;
       return;
     }
 
-    re2::StringPiece prefix(inputStringPiece.data(), utf8Position);
-    re2::StringPiece targetStringPiece(
-        inputStringPiece.data() + utf8Position,
-        inputStringPiece.size() - utf8Position);
-
+    std::string prefix(stringInput.data(), utf8Position);
     std::string targetString(
-        targetStringPiece.data(), targetStringPiece.size());
-    RE2::GlobalReplace(&targetString, *patternRegex, replaceStringPiece);
+        stringInput.data() + utf8Position, stringInput.size() - utf8Position);
 
-    if (targetString.size() || prefix.size()) {
-      result.resize(prefix.size() + targetString.size());
-      std::memcpy(result.data(), prefix.data(), prefix.size());
-      std::memcpy(
-          result.data() + prefix.size(),
-          targetString.data(),
-          targetString.size());
-    } else {
-      result.resize(0);
-    }
+    RE2::GlobalReplace(&targetString, *patternRegex, replaceStringPiece);
+    result = prefix + targetString;
   }
 
  private:
-  re2::RE2* getCachedRegex(const std::string& pattern) const {
-    auto it = patternCache_.find(pattern);
-    if (it != patternCache_.end()) {
+  re2::RE2* getRegex(const std::string& pattern) const {
+    auto it = cache_.find(pattern);
+    if (it != cache_.end()) {
       return it->second.get();
     }
     VELOX_USER_CHECK_LT(
-        patternCache_.size(),
+        cache_.size(),
         kMaxCompiledRegexes,
         "regexp_replace hit the maximum number of unique regexes: {}",
         kMaxCompiledRegexes);
     checkForCompatiblePattern(pattern, "regexp_replace");
     auto patternRegex = std::make_unique<re2::RE2>(pattern);
     auto* rawPatternRegex = patternRegex.get();
     checkForBadPattern(*rawPatternRegex);
-    patternCache_.emplace(pattern, std::move(patternRegex));
+    cache_.emplace(pattern, std::move(patternRegex));
     return rawPatternRegex;
   }
 
-  mutable folly::F14FastMap<std::string, std::unique_ptr<re2::RE2>>
-      patternCache_;
+  mutable folly::F14FastMap<std::string, std::unique_ptr<re2::RE2>> cache_;
+
+  size_t getUTFLength(const char* str, size_t len, size_t max_length) {
+    // Adjust the position for UTF-8 by counting the code points.
+    size_t utf8Position = 0;
+    size_t numCodePoints = 0;
+    while (numCodePoints < max_length - 1 && utf8Position <= len) {
```

Comment:
If I am reading that correctly, I believe so. I would call it as follows

```cpp
size_t utf8Position = 
facebook::velox::functions::stringImpl::cappedLength::cappedLength<false>(stringInput, position);
```

I have two concerns and one idea for readability:

1. My implementation throws when it encounters a bad byte, cappedLengthUnicode skips it. This should cause only regexpReplaceUnvalidUTF8 to fail in testing.
2. My implementation is aware of the '1-indexed' condition, whereas cappedLength appears to be 0-indexed. 
3. The call has a pretty long namespace. For readability, just make getUTFLength an encapsulation of cappedLength so the code is still clear about what it is doing. If you feel this is unnecessary, I can just make the full call in a single line.

While writing this up I built and tested, the following test cases now fail:

```
[  FAILED  ] 16 tests, listed below:
[  FAILED  ] RegexFunctionsTest.regexpReplaceRegistration
[  FAILED  ] RegexFunctionsTest.regexpReplaceNonAsciiPositionNoChange
[  FAILED  ] RegexFunctionsTest.regexpReplaceMatchSparkSqlTestSimple
[  FAILED  ] RegexFunctionsTest.regexpReplaceWithEmptyString
[  FAILED  ] RegexFunctionsTest.regexpReplaceInvalidUTF8
[  FAILED  ] RegexFunctionsTest.regexpReplacePosition
[  FAILED  ] RegexFunctionsTest.regexpReplaceMultipleMatches
[  FAILED  ] RegexFunctionsTest.regexpReplaceDataframe
[  FAILED  ] RegexFunctionsTest.regexpReplaceDataframeMultiple
[  FAILED  ] RegexFunctionsTest.regexpReplaceDataframeSpecial
[  FAILED  ] RegexFunctionsTest.regexpReplaceDataframeSizes
[  FAILED  ] RegexFunctionsTest.regexpReplaceDataframeCharacterTypes
[  FAILED  ] RegexFunctionsTest.regexpReplaceMatchSparkSqlTest
[  FAILED  ] RegexFunctionsTest.regexpReplaceRowsNoPosition
[  FAILED  ] RegexFunctionsTest.regexpReplaceMassiveVectors
[  FAILED  ] RegexFunctionsTest.regexpReplaceCacheMissLimit

```
Passing position - 1 and adding 1 to the output does not resolve any test cases, in fact an additional test case fails. This isn't necessarily because of the new function, it could be how I am calling it, so I will take a deeper look and gather details.

Reply:
Adjusting the call to use "position - 1" and not add 1 back to the end was the correct call. It now only fails the following:

```
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] RegexFunctionsTest.regexpReplaceNonAsciiPositionNoChange
[  FAILED  ] RegexFunctionsTest.regexpReplaceInvalidUTF8
```

I'll look into how to fix the first since we can just remove or adjust the second. 

Is skipping bad bytes the standard way of handling them? I'm re-rewriting the InvalidUTF8 case to instead expect some sort of valid output, but I am not sure if this is the right idea.

Reply:
@mbasmanova I don't think we can replace getUTF8Positon with cappedLength. 

cappedLength returns the number of characters it encounters:
```cpp
FOLLY_ALWAYS_INLINE int64_t
cappedLengthUnicode(const char* input, size_t size, size_t maxChars) {
  // First address after the last byte in the input
  auto end = input + size;
  auto currentChar = input;
  int64_t numChars = 0;

  // Use maxChars to early stop to avoid calculating the whole
  // length of long string.
  while (currentChar < end && numChars < maxChars) {
    auto charSize = utf8proc_char_length(currentChar);
    // Skip bad byte if we get utf length < 0.
    currentChar += UNLIKELY(charSize < 0) ? 1 : charSize;
    numChars++;
  }

  return numChars;
}
```

whereas getUTF8Length returns the sum of the number of bytes it has encountered

```cpp

  size_t getUTFLength(const char* str, size_t len, size_t maxCharacters) {
    // Adjust the position for UTF-8 by counting the code points.
    size_t utf8Position = 0;
    size_t numCodePoints = 0;
    int64_t numChars = 0;
    auto currentChar = str;
    auto end = str + len;
    while (numCodePoints < maxCharacters - 1 && utf8Position <= len) {
      int charLength = utf8proc_char_length(str + utf8Position);
      VELOX_USER_CHECK_GT(
          charLength, 0, "regexp_replace encountered invalid UTF-8 character");
      ++numCodePoints;
      utf8Position += charLength;
    }
```

So, if we encounter multi-byte characters like the 3 byte ♫ or the 2 byte ¡, the return values differ between the two functions.

getUTFLength's return value is used to split the string between prefix and target, so those extra bytes make a difference in some cases.